### PR TITLE
Handle symlinks while uploading files in directory

### PR DESCRIPTION
**Pull request by ysaito1001**
~~**Will update this PR, since with the first commit, the default case (not following symlinks) has to pay the price of getting metadata twice which is inefficient**~~

*Description of changes:*
This PR is a follow-up on #67 and handles uploading files with symbolic links.

The primary change is to use `fs::symbolink_metadata` to property detect whether a given path entry represents a symbolic link. Unlike `fs::metadata`, which follows symbolic links and returns metadata about the target, `fs::symlink_metadata` retrieves metadata about the symlink itself.

Also added `test_source_dir_is_symlink`, adapted from the [Java SDK V2
](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L398-L413)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-s3-transfer-manager/src/types.rs [hunk 1]
```diff
@@ -179,7 +179,20 @@ fn is_hidden(path: &Path) -> bool {
 impl Default for UploadFilter {
     fn default() -> Self {
         Self {
-            predicate: Arc::new(|item| item.metadata().is_file() && !is_hidden(item.path())),
+            predicate: Arc::new(|item| {
+                if is_hidden(item.path()) {
+                    return false;
+                }
+                if item.symlink_metadata.is_symlink() {
```
Comment (graebm) [inline aws-s3-transfer-manager/src/types.rs:186]: If our transfer manager's `follow_symlinks` setting is `false`, should it stop us from uploading symlinked files? Or is only supposed to stop us from recursing *into* symbolic directories?

If it's supposed to stop us from uploading symbolic files, then we shouldn't need to have this check here in the `UploadFilter`. Our code should have rejected this DirEntry before presenting it to the user's `UploadFilter`

I tested `WalkDir` and its [WalkDir.follow_links(false)](https://docs.rs/walkdir/latest/walkdir/struct.WalkDir.html#method.follow_links) only stops it from *entering inside* symbolic directories. It will still yield the symbolic directories and files as it iterates.

File: aws-s3-transfer-manager/src/types.rs [hunk 2]
```diff
@@ -189,28 +202,84 @@ impl Default for UploadFilter {
 #[derive(Debug)]
 pub struct UploadFilterItem<'a> {
     pub(crate) path: Cow<'a, Path>,
-    pub(crate) metadata: Metadata,
+    pub(crate) symlink_metadata: Metadata,
+    pub(crate) metadata: Option<Metadata>,
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:206]: I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, that's ok since it does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:206]: I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, `metadata`, the one knows the file type of the symlink target, can be `None` since the upload operation does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

File: aws-s3-transfer-manager/src/types.rs (hunk 1 above)
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:186]: > should it stop us from uploading symlinked files? Or is only supposed to stop us from recursing into symbolic directories?

Good question. The SEP doesn't seem to specify this behavior, but according to a [Java SDK V2's test](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L248-L272), it's the former (symlinked files are [not in the expected test result](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L271)) and so is the implementation in this PR.

> If it's supposed to stop us from uploading symbolic files, then we shouldn't need to have this check here in the UploadFilter

We actually need this check, because a symlink can be a link to a single file, in which case, a `DirEntry` representing the symlink to the single file is still presented to `UploadFilter`. In fact, if we removed `if item.symlink_metadata.is_symlink()`, I believe [this unit test](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/08330ab0bb1fb4d9a4c62595d2525d2f0e2ca8e9/aws-s3-transfer-manager/tests/upload_objects_test.rs#L204-L218) would fail because `output.objects_uploaded()` would become 2, including `symlink2` as a file to upload as well.

Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:186]: > should it stop us from uploading symlinked files? Or is only supposed to stop us from recursing into symbolic directories?

Good question. The SEP doesn't seem to specify this behavior, but according to a [Java SDK V2's test](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L248-L272), it's the former (symlinked files are [not in the expected test result](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L271)) and so is the implementation in this PR.

> If it's supposed to stop us from uploading symbolic files, then we shouldn't need to have this check here in the UploadFilter

I'm pretty sure we need this check, because a symlink can be a link to a single file, in which case, a `DirEntry` representing the symlink to the single file is still presented to `UploadFilter`. If we removed `if item.symlink_metadata.is_symlink()` and only left `item.symlink_metadata.is_file()` as the last expression, I believe [this unit test](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/08330ab0bb1fb4d9a4c62595d2525d2f0e2ca8e9/aws-s3-transfer-manager/tests/upload_objects_test.rs#L186-L202) would fail because `output.objects_uploaded()` would become 6, incorrectly rejecting `symlink2` as `item.symlink_metadata.is_file()` returns false (and `item.symlink_metadata.is_symlink()` returns true).

File: aws-s3-transfer-manager/src/types.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:206]: I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, `metadata`, the one knows the file type of the symlink target, can be `None` since the upload operation does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

In short, `metadata`, which follows symbolic links, is not suitable for the upload operation's default case where it does not follow symbolic links, so it becomes `Some` only when the upload operation is configured to follow symbolic links.

[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:206]: I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, `metadata`, the one knows the file type of the symlink target, can be `None` since the upload operation does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

In short, `metadata`, which follows symbolic links, is not suitable for the upload operation's default case where it does not follow symbolic links. It is an `Option` because we don't want to pay the additional price of calling `fs::metadata` if we don't need it; we already call `fs::symlink_metadata` to work with upload operation's default case that does not follow symbolic links, and `Metadata` returned by `fs::symlink_metadata` can still answer whether a path entry is a file or a directory if the path entry is not a symbolic link.

File: aws-s3-transfer-manager/src/operation/upload_objects.rs [hunk 3]
```diff
@@ -37,7 +37,15 @@ impl UploadObjects {
     ) -> Result<UploadObjectsHandle, crate::error::Error> {
         //  validate existence of source and return error if it's not a directory
         let source = input.source().expect("source set");
-        validate_target_is_dir(source).await?;
+        let metadata = tokio::fs::symlink_metadata(source).await?;
+        validate_target_is_dir(source).await.and_then(|_| {
+            if !input.follow_symlinks() && metadata.is_symlink() {
+                return Err(error::invalid_input(
+                    format!("{source:?} is a symbolic link to a directory but the current upload operation does not follow symbolic links"))
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects.rs:44]: That's exactly the tests I called out in the PR description:

> Also added test_source_dir_is_symlink, adapted from the [Java SDK V2](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L398-L413)

File: aws-s3-transfer-manager/src/operation/upload_objects.rs [hunk 4]
```diff
@@ -37,7 +37,15 @@ impl UploadObjects {
     ) -> Result<UploadObjectsHandle, crate::error::Error> {
         //  validate existence of source and return error if it's not a directory
         let source = input.source().expect("source set");
-        validate_target_is_dir(source).await?;
+        let metadata = tokio::fs::symlink_metadata(source).await?;
+        validate_target_is_dir(source).await.and_then(|_| {
+            if !input.follow_symlinks() && metadata.is_symlink() {
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects.rs:42]: Personally, I see this as a wrinkle in the type system brought by [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html), because we cannot tell how `Metadata` was originally obtained. When `fs::metadata` is used, the returned `Metadata` can only provide information about whether a path is a file or a directory, but it cannot tell if the path is a symlink. On the other hand, `fs::symlink_metadata` (which [corresponds to `lstat` on Unix](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html)) can identify whether a path is a symlink. However, if the path is a symlink, the `Metadata` returned by `fs::symlink_metadata` can only indicate that it is a symlink via `is_symlink()`. It will return false for both `is_file()` and `is_dir()`, even though the symlink may point to a file or directory.

> Can just use call metadata.is_dir() since we already looked up the metadata

Yes we can. For the reason above, we have to obtain `Metadata` twice in the worst case but for the default case where the upload operation does not follow symbolic links, we should obtain it only once (I overlooked this place as part commit 08330ab).

File: aws-s3-transfer-manager/src/operation/upload_objects.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects.rs:44]: > What does Java Transfer Manager do here?

That's exactly the tests I called out in the PR description:

> Also added test_source_dir_is_symlink, adapted from the [Java SDK V2](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L398-L413)

File: aws-s3-transfer-manager/src/operation/upload_objects.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects.rs:42]: Personally, I see this as a wrinkle in the type system brought by [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html), because we cannot tell how `Metadata` was originally obtained (but maybe by design). When `fs::metadata` is used, the returned `Metadata` can only provide information about whether a path is a file or a directory, but it cannot tell if the path is a symlink. On the other hand, `fs::symlink_metadata` (which [corresponds to `lstat` on Unix](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html)) can identify whether a path is a symlink. However, if the path is a symlink, the `Metadata` returned by `fs::symlink_metadata` can only indicate that it is a symlink via `is_symlink()`. It will return false for both `is_file()` and `is_dir()`, even though the symlink may point to a file or directory.

> Can just use call metadata.is_dir() since we already looked up the metadata

Yes we can. For the reason above, we have to obtain `Metadata` twice in the worst case but for the default case where the upload operation does not follow symbolic links, we should obtain it only once (I overlooked this place as part of commit 08330ab).

[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects.rs]: e17439f obtains `Metadata` only once for the common case during validation.

File: aws-s3-transfer-manager/src/operation/upload_objects/worker.rs [hunk 5]
```diff
@@ -423,6 +449,87 @@ mod unit {
...
+                ("docs/2022/February/sample1.txt", 1),
+                ("docs/2022/February/sample2.txt", 1),
+                ("docs/2022/February/sample3.txt", 1),
+            ];
+            let temp_dir2 = create_test_dir(Some("temp2"), files2.clone(), &[]);
+
+            let files3 = vec![("sample3.png", 1)];
+            let temp_dir3 = create_test_dir(Some("temp3"), files3.clone(), &[]);
+
+            // Crate a symbolic link from `temp1/symlink` to `temp2`
```
Comment (graebm) [inline aws-s3-transfer-manager/src/operation/upload_objects/worker.rs:469]: trivial:
```suggestion
            // Create a symbolic link from `temp1/symlink` to `temp2`
```

File: aws-s3-transfer-manager/src/operation/upload_objects/worker.rs [hunk 6]
```diff
@@ -423,6 +449,87 @@ mod unit {
...
+            symlink(&temp_dir2, temp_dir1.path().join("symlink"))
+                .await
+                .unwrap();
+            // Crate a symbolic link from `temp1/symlink2` to `temp3/sample.png`
```
Comment (graebm) [inline aws-s3-transfer-manager/src/operation/upload_objects/worker.rs:473]: trivial
```suggestion
            // Create a symbolic link from `temp1/symlink2` to `temp3/sample.png`
```

File: aws-s3-transfer-manager/src/types.rs (hunk 1 above)
Comment (graebm) [reply aws-s3-transfer-manager/src/types.rs:186]: I get that the `is_symlink()` check needs to happen. But I'm saying the Transfer Manager should be checking this *before* presenting the item to the user-written `UploadFilter`.

Since `WalkDir` is going to present the Transfer manager with symlinked files even if when `follow_symlinks` is false, it's up to the Transfer Manager to skip over these without relying on the user's `UploadFilter`.

Comment (graebm) [reply aws-s3-transfer-manager/src/types.rs:186]: And if we're not relying on the `UploadFilter` to deal with symlinks, then we could simplify `UploadFilterItem` to only have the actual `metadata` (vs having both `symlink_metadata` and optional `metadata`)

Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs:186]: Ah, got it. Thanks for clarifying. Yes, that should help simplify things. I'll make that adjustment.

File: aws-s3-transfer-manager/src/types.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs]: `symlink_metadata` got removed and `metadata` is no longer an optional in 5a52119.

**Review by aajtodd — APPROVED** [review]

**Review by graebm — APPROVED** [review]

File: aws-s3-transfer-manager/src/operation/upload_objects/worker.rs [hunk 7]
```diff
@@ -51,8 +51,23 @@ pub(super) async fn list_directory_contents(
...
+                } else {
+                    // In this branch, we know `symlink_metadata` does not represent a symlink link,
+                    // so it can return true for either `is_dir()` or `is_file()`.
+                    symlink_metadata
+                };
+                let filter_item = UploadFilterItem::builder()
+                    .path(entry.path())
+                    .metadata(metadata.clone())
+                    .build();
+                if !(filter.predicate)(&filter_item) {
```
Comment (graebm) [inline aws-s3-transfer-manager/src/operation/upload_objects/worker.rs:70]: Are we doing anything different in this loop based on whether `entry` is a file or directory? If so, I don't see where

File: aws-s3-transfer-manager/src/types.rs [hunk 8]
```diff
@@ -179,7 +179,20 @@ fn is_hidden(path: &Path) -> bool {
 impl Default for UploadFilter {
     fn default() -> Self {
         Self {
-            predicate: Arc::new(|item| item.metadata().is_file() && !is_hidden(item.path())),
+            predicate: Arc::new(|item| {
+                if is_hidden(item.path()) {
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs]: Good at spotting a hole in the logic :)

> If a user purposefully sets source = "/path/to/.hidden/subfolder_to_upload"

Wrote a dummy unit test and confirmed `subfolder_to_upload` will be yielded as a file to upload to S3. Also confirmed that Java Transfer Manager only supports [`DownloadFilter`](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/config/DownloadFilter.java#L30) but not `UploadFilter` currently.

At least I'll restore your original [TODO](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/c88e07f7c92a3fe8bd391c624876538743c5314b/aws-s3-transfer-manager/src/operation/upload_objects/builders.rs#L110-L115), move it to `UploadFilter::default()`, and update it to say it does not filter a hidden directory.

> Would it make more sense if we called the UploadFilter on directories to decide whether to enter them?

Yes it would, and it'd be ideal if `WalkDir` could support that out of the box; even if we successfully apply the `UploadFilter` to skip a hidden directory, subsequent calls to `WalkDir::next()` will still yield files from within that hidden directory, so if we could stop the iteration at the source, that'd be great.

If we need to filter out hidden directories in `UploadFilter`, we should avoid repeatedly inspecting every path component of a `DirEntry` to check if it's inside a hidden directory; doing so in each `UploadFilter ` call could negatively impact the performance of the common case. Maintaining a hash set of hidden directories can be tricky as well, especially if we want to parallelize directory traversal for future performance improvements, as we might have to give up the guarantee of directory traversal order.

[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/types.rs]: Good at spotting a hole in the logic :)

> If a user purposefully sets source = "/path/to/.hidden/subfolder_to_upload"

Wrote a dummy unit test and confirmed `subfolder_to_upload` will be yielded as a file to upload to S3. Also confirmed that Java Transfer Manager only supports [`DownloadFilter`](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/config/DownloadFilter.java#L30) but not `UploadFilter` currently.

At least I'll add a comment to `UploadFilter::default()` that says it does not filter hidden directories

> Would it make more sense if we called the UploadFilter on directories to decide whether to enter them?

Yes it would, and it'd be ideal if `WalkDir` could support that out of the box; even if we successfully apply the `UploadFilter` to skip a hidden directory, subsequent calls to `WalkDir::next()` will still yield files from within that hidden directory, so if we could stop the iteration at the source, that'd be great.

If we need to filter out hidden directories in `UploadFilter`, we should avoid repeatedly inspecting every path component of a `DirEntry` to check if it's inside a hidden directory; doing so in each `UploadFilter ` call could negatively impact the performance of the common case. Maintaining a hash set of hidden directories can be tricky as well, especially if we want to parallelize directory traversal for future performance improvements, as we might have to give up the guarantee of directory traversal order.

File: aws-s3-transfer-manager/src/operation/upload_objects/worker.rs (hunk 7 above)
Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects/worker.rs:70]: No we're not. Both types, a directory and a file, are presented to the filter.

Comment (ysaito1001) [reply aws-s3-transfer-manager/src/operation/upload_objects/worker.rs:70]: No we're not. Both types, a directory and a file, are yielded in the loop. But the default filter only includes files.
